### PR TITLE
EVM loader program support via SputnikVM

### PR DESCRIPTION
**Pull request by sorpaas**
#### Problem

This is a draft PR for proof of concept EVM loader support for Solana, via SputnikVM. Can't say I really understand how the state transition function works in Solana, so please let me know if I got something terribly wrong!

#### Summary of Changes

This adds a Solana program that has a `process_instruction` function:

* `LoadInstruction::Write` is expected to write EVM init code into a particular account.
* `LoadInstruction::Finalize` will then execute the init code. Subsequent keyed accounts are expected to be all state accesses of that execution. If the execution is successful, the account will be marked as executable, and its data will be swapped to an encoding of `(Address, EthereumAccount)`.
* `LoadInstruction::InvokeMain` will execute a contract.

Known issues:

* TODO: definitely more known issues than what's listed below.
* This is mostly a question, probably quite dumb -- where can I find descriptions on the lifecycle of accounts? In particular, how would a program create new accounts? I saw `system_instruction::create_account` but that only returns an `Instruction`?
* All storage data of an account are stored in one Solana account's `data` field, which means all of them must be encoded/decoded when accessed.
* We currently use `MemoryBackend`, which duplicates all data in `KeyedAccount` in memory again. This should be changed to a custom backend for performance.

### CODE REVIEW

**Comment by mvines** [comment]
@sorpaas this looks like a great start, @jackcmay can probably help answer some of your questions.   You might want to find us on Discord as well so we can chat more interactively as needed

**Comment by sakridge** [comment]
> This is mostly a question, probably quite dumb -- where can I find descriptions on the lifecycle of accounts? In particular, how would a program create new accounts? I saw system_instruction::create_account but that only returns an Instruction?

Yes, a Transaction contains a list of Instructions, so a single Transaction could have 1 CreateAccount Instruction or be bundled with a CreateAccount and something else like an invoke of another smart contract. I don't believe we have a way to actually create a new account from within the contract code.

**Comment by jackcmay** [comment]
> > This is mostly a question, probably quite dumb -- where can I find descriptions on the lifecycle of accounts? In particular, how would a program create new accounts? I saw system_instruction::create_account but that only returns an Instruction?
> 
> Yes, a Transaction contains a list of Instructions, so a single Transaction could have 1 CreateAccount Instruction or be bundled with a CreateAccount and something else like an invoke of another smart contract. I don't believe we have a way to actually create a new account from within the contract code.

Here is an example of another programing doing what @sakridge is describing.  Creating an account and then using that account in its `initialize` instruction:
https://github.com/solana-labs/solana/blob/43e608af477a8c7725cf55ed606af7f658330cd2/programs/stake/src/stake_instruction.rs#L143

File: programs/evm_loader/src/processor.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            origin: H160::default(),
+            chain_id: U256::zero(),
+            block_hashes: Vec::new(),
+            block_number: U256::zero(),
+            block_coinbase: H160::default(),
+            block_timestamp: U256::zero(),
+            block_difficulty: U256::zero(),
+            block_gas_limit: U256::zero(),
+        };
+        let code = keyed_account.account.data.clone();
```
Comment (jackcmay) [inline programs/evm_loader/src/processor.rs]: Why clone here?

File: programs/evm_loader/src/processor.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+                },
+                Apply::Delete {
+                    address
+                } => {
+                    let mut processed = false;
+                    'inner_delete: for state_account in keyed_accounts.iter_mut() {
+                        let (keyed_address, _) =
+                            Self::read_account(&state_account.account.data)?;
+
+                        if keyed_address == address {
```
Comment (jackcmay) [inline programs/evm_loader/src/processor.rs]: nit: Current variable names might be misleaading.  These `keyed_address` and `keyed_account` types are not Solana `KeyedAccount` types but `H160` and `MemoryAccount`?

File: programs/evm_loader/src/processor.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        let exit_reason =
+            executor.transact_create(keyed_address, U256::zero(), code, usize::max_value());
+
+        let (applies, _logs) = executor.deconstruct();
+
+        if exit_reason.is_succeed() {
+            keyed_account.account.executable = true;
+        }
+
+        Self::apply_accounts(applies, keyed_accounts)?;
```
Comment (jackcmay) [inline programs/evm_loader/src/processor.rs]: What state updates are we expecting here?

File: programs/evm_loader/src/processor.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        let backend = MemoryBackend::new(&vicinity, state);
+        let config = evm::Config::istanbul();
+        let mut executor = StackExecutor::new(&backend, usize::max_value(), &config);
+        let exit_reason =
+            executor.transact_create(keyed_address, U256::zero(), code, usize::max_value());
+
+        let (applies, _logs) = executor.deconstruct();
+
+        if exit_reason.is_succeed() {
+            keyed_account.account.executable = true;
```
Comment (jackcmay) [inline programs/evm_loader/src/processor.rs]: If there is any other verification of the program that we can do it should be done before its marked `executable`.  The goal is for `do_invoke_main` to only fail due to inputs to the finalized program and not due to the finalized program itself

File: programs/evm_loader/src/processor.rs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                    reset_storage,
+                } => {
+                    let mut processed = false;
+                    'inner_modify: for state_account in keyed_accounts.iter_mut() {
+                        let (keyed_address, mut keyed_account) =
+                            Self::read_account(&state_account.account.data)?;
+
+                        if keyed_address == address {
+                            keyed_account.nonce = basic.nonce;
+                            keyed_account.balance = basic.balance;
```
Comment (jackcmay) [inline programs/evm_loader/src/processor.rs]: What does this balance represent?

**Pull request closed without merging**